templates/create-daml-app-test-resources/index.test.ts

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { ChildProcess, spawn, SpawnOptions } from 'child_process';
+import waitOn from 'wait-on';
+
+import Ledger from '@daml/ledger';
+import { User } from '@daml.js/create-daml-app';
+import { computeCredentials } from './Credentials';
+
+import puppeteer, { Browser, Page } from 'puppeteer';
+
+const SANDBOX_PORT = 6865;
+const JSON_API_PORT = 7575;
+const UI_PORT = 3000;
+
+// `daml start` process (which spawns a sandbox and JSON API server)
+let startProc: ChildProcess | undefined = undefined;
+
+// Headless Chrome browser:
+// https://developers.google.com/web/updates/2017/04/headless-chrome
+let browser: Browser | undefined = undefined;
+
+let uiProc: ChildProcess | undefined = undefined;
+
+// Function to generate unique party names for us.
+// This should be replaced by the party management service once that is exposed
+// in the HTTP JSON API.
+let nextPartyId = 1;
+function getParty(): string {
+  const party = `P${nextPartyId}`;
+  nextPartyId++;
+  return party;
+}
+
+test('Party names are unique', async () => {
+  const parties = new Set(Array(10).fill({}).map(() => getParty()));
+  expect(parties.size).toEqual(10);
+});
+
+// Use a single sandbox, JSON API server and browser for all tests for speed.
+// This means we need to use a different set of parties and a new browser page for each test.
+beforeAll(async () => {
+  // Run `daml start` to start up the sandbox and json api server.
+  // Run it from the repository root, where the `daml.yaml` lives.
+  // The path should already include '.daml/bin' in the environment where this is run.
+  const startOpts: SpawnOptions = { cwd: '..', stdio: 'inherit' };
+  startProc = spawn('daml', ['start'], startOpts);
+
+  // Run `yarn start` in another shell.
+  // Disable automatically opening a browser using the env var described here:
+  // https://github.com/facebook/create-react-app/issues/873#issuecomment-266318338
+  const env = {...process.env, BROWSER: 'none'};
+  uiProc = spawn('yarn', ['start'], { env, stdio: 'inherit', detached: true});
+  // Note(kill-yarn-start): The `detached` flag starts the process in a new process group.
+  // This allows us to kill the process with all its descendents after the tests finish,
+  // following https://azimi.me/2014/12/31/kill-child_process-node-js.html.
+
+  // We know the `daml start` and `yarn start` servers are ready once the relevant ports become available.
+  await waitOn({resources: [
+    `tcp:localhost:${SANDBOX_PORT}`,
+    `tcp:localhost:${JSON_API_PORT}`,
+    `tcp:localhost:${UI_PORT}`
```

Comment:
In a follow-up PR, let's wait for the `--port-file` to be created for sandbox and the JSON API instead of waiting for the ports, which has been less reliable, at least on MacOS, in the past. Let's do this regardless of whether we switch to dynamic port allocation.